enhancements/network/ocp-on-smart-nics.md

```diff
@@ -0,0 +1,313 @@
+containing a Smart NIC must have offloaded networking, which, in the
+current architecture, means it must explicitly _request_ offloaded
+networking. In particular, this means that most pods created by
+OpenShift operators cannot be deployed on a node with a Smart NIC, and
+so all masters and at least a few workers will need to have "dumb"
+NICs.
+
+(While the exact architectural constraints may change in the future,
+it is likely that some customers will always want the possibility of
+only having Smart NICs on a subset of nodes.)
+
+### Overall Cluster Architecture
+
+There are several possible overall architectures:
+
+  - The Smart NICs could be nodes within the same OCP cluster as their
+    hosts. We would need to use some combination of labels and taints
+    to distinguish "real" nodes from Smart NIC nodes. (Among other
+    things, the Smart NICs would be their own `MachineConfigPool`,
+    separate from the masters and workers.) This may require changes
+    to tolerations and node selectors on existing DaemonSets, both
+    within OCP and in third-party operators or customer workloads (and
+    it may turn out that this approach is unworkable for this reason).
+
+  - The Smart NICs could be part of a second cluster running in
+    parallel with the primary cluster, with the control plane of the
+    NIC cluster running as pods within the primary cluster.
+
+  - The Smart NICs could be part of a second cluster which is
+    completely separate from the primary cluster, with its own master
+    nodes. (In some external discussion of this approach, it was
+    assumed that the masters in this scenario would have to be on ARM
+    hosts, but this isn't required; you can have a mix of x86_64
+    master nodes and ARM Smart NIC "worker" nodes.)
+
+  - The Smart NICs could be part of a second cluster, where the
+    control planes for both clusters are running in a different
+    cluster, ROKS-style.
+
+The latter two approaches allow for more isolation between the hosts
+and the NICs, since potentially even the cluster administrators of the
+primary cluster might not be able to influence the NIC cluster. This
+would be especially important if additional primary-cluster security
+and monitoring functions were offloaded to the NICs in the future.
+
+### Initial Development With Beta BlueField-2 Hardware
+
+[NOTE: We are confirming the timing/availability of various hardware.
+It may turn out that it doesn't make sense for us to worry about these
+cards.]
+
+Although we will eventually want to use IPMI to control and provision
+the NICs, just like we do for the hosts with bare-metal IPI, this will
+not work with the pre-GA BlueField-2 cards that we currently have
+access to. Instead, we will have to communicate with the NICs over PCI
+from the host. The most future-compatible approach is probably to
+create an IPMI proxy that can run on each host, claiming its own
+separate IP address on the provisioning network, and then proxying
```

Comment:
Is this `provisioning network` dedicated to smart NIC provisioning?

Reply:
No, there's already a provisioning network.

For non-BlueField bare-metal installs, the expectation is that your host has a spiffy PCI NIC that does 10-100G ethernet and offload and stuff, which is what's used for real traffic, but there's also another wimpier ethernet port soldered to the motherboard that lets you do IPMI, etc, to manage the host. So we install the node via that management port, and then transition to using the "real" NIC for everything else after that. The assumption is that for security reasons they'd be connected to different ethernet networks, though that isn't actually required.

For real (future) BlueField installs, the expectation is the same, except that in addition to the management port on the host motherboard, there's _another_ management port on the BF-2, so you'll have three ethernet cables: one for host management/provisioning, one for NIC management/provisioning, and one for real cluster traffic. (Or maybe two for traffic; there are actually 2 100G ports on the BF-2 in addition to the 1G management port.) I imagine the host management port and the NIC management port would be connected to the same network as each other (which is still separate from the real network), because otherwise you'd need two sets of PXE servers and stuff...

For the "IPMI proxy" scenario, we can't use the management port on the BF-2, so instead we have the host claim a second IP address on its own management port's network, and pretend that that's the BF-2 management port.